xwiki-platform-core/xwiki-platform-security/xwiki-platform-security-requiredrights/xwiki-platform-security-requiredrights-ui/src/main/java/org/xwiki/platform/security/requiredrights/ui/MissingRequiredRightWarningUIExtension.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.platform.security.requiredrights.ui;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import javax.annotation.Priority;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.slf4j.Logger;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.platform.security.requiredrights.RequiredRightAnalysisResult;
+import org.xwiki.platform.security.requiredrights.RequiredRightAnalyzer;
+import org.xwiki.platform.security.requiredrights.internal.RequiredRightChangeSuggestion;
+import org.xwiki.platform.security.requiredrights.internal.RequiredRightsChangeSuggestionManager;
+import org.xwiki.rendering.block.Block;
+import org.xwiki.rendering.block.CompositeBlock;
+import org.xwiki.rendering.block.GroupBlock;
+import org.xwiki.security.authorization.AuthorizationManager;
+import org.xwiki.security.authorization.Right;
+import org.xwiki.security.authorization.requiredrights.DocumentRequiredRights;
+import org.xwiki.security.authorization.requiredrights.DocumentRequiredRightsManager;
+import org.xwiki.skinx.SkinExtension;
+import org.xwiki.template.TemplateManager;
+import org.xwiki.uiextension.UIExtension;
+
+import com.xpn.xwiki.XWikiContext;
+
+/**
+ * Displays a warning above the content if required rights are missing.
+ *
+ * @version $Id$
+ */
+@Component
+@Named(MissingRequiredRightWarningUIExtension.ROLE_HINT)
+@Priority(100)
```

Comment:
I would document that priority.

Reply:
Good point, I'm actually not sure why I chose that particular priority… I think I'll actually increase it to be above the default of 1000, maybe set it to 2000 such that it is displayed closer to the content. What do you think? From what I could find, the relations application is the only (public) code using [this UIXP](https://www.xwiki.org/xwiki/bin/view/Documentation/DevGuide/ExtensionPoint/AfterContentHeaderUIX).

Reply:
You actually have a few contrib extensions using it: e.g. replication or book version AFAICS; https://github.com/search?q=org%3Axwiki-contrib%20org.xwiki.platform.template.content.header.after&type=code

Reply:
Indeed, I must have remembered this incorrectly. Now I think none of them define a priority as the UIXP doesn't support any parameters, and we don't support specifying the component priority in UIX defined in wiki pages. And the one defined in a Java component also doesn't specify an order. So basically the question is: should it before or after all those UIX. Do you have any opinions/do you know if any of them has any requirements?

Reply:
> as the UIXP doesn't support any parameters

We should probably fix that to introduce a parameter order no? Would be probably cleaner. 

> Do you have any opinions/do you know if any of them has any requirements?

Well I agree with you that your UIXP should probably be closer to the content than any other currently existing.

Reply:
Honestly I don't have a strong opinion so either way is fine with me.